### PR TITLE
Added classify method.

### PR DESCRIPTION
**Pull request by primaryobjects**
Added classify method: returns the index of a cluster for a new data point.
Usage:
var clusters = clusterfck.kmeans(colors, 3);
var clusterIndex = clusterfck.kmeans.classify([0, 0, 225]);

Output:
1

### CODE REVIEW

**Comment by irony** [comment]
Great work! This is exactly what I'm looking for. Shouldnt the centeroids array be public so it can be serialized and used later when new data should be clustered?

**Comment by primaryobjects** [comment]
Actually, I just pushed changes that include toJSON() and fromJSON() methods for serializing the centroids. See https://github.com/primaryobjects/clusterfck

The return value from the kmeans() method now returns an object containing { clusters, centroids }.

I've also added K-means++ initialization of the centroids (as opposed to random initialization).

**Comment by irony** [comment]
Great! I will try it out. Maybe to not break backward compatibility it would be better to add the centeroids as a property on the array instead of creating a new object? Like this:

```
var clusters = clusterfck.kmeans(vectors);
console.log(clusters.centroids); 
```

Anyway, good work and perfect timing, I have been using this lib for a while and was just going to expand the lib with this functionality - two days before I did, you pushed the same changes ;)

**Comment by primaryobjects** [comment]
Hey, give this branch a try. I've converted the functions to a class with prototype methods, so you can now create multiple instances. If it looks good, I'll move it to the master of my fork.
https://github.com/primaryobjects/clusterfck/tree/PROTOTYPE

The new usage would be:
var clusterfck = require('clusterfck');

var kmeans = new clusterfck.kmeans();
var clusters = kmeans.train(colors, 3);
var clusterIndex = kmeans.classify(input);

**Comment by irony** [comment]
Cool, i'll let you know!

**Comment by irony** [comment]
It works fine! I suggest adding this in the constructor?

```
function KMeans(centroids) {
this.centroids = centroids || [];
}
```

That means you can do this:

```
var kmeans = new clusterfck.kmeans(centroids);
```

**Comment by primaryobjects** [comment]
Done. I've added the optional constructor parameter. Also, renamed train() to cluster() since it makes more sense. 
https://github.com/primaryobjects/clusterfck/commit/9e5cdbdd2e648cf0f2c93f2d808beeec18e7a425

Now, if we can just get harthur to accept the pull request and update the npm package! :)

**Comment by irony** [comment]
Great work! Pull it @harthur :+1:

**Comment by harthur** [comment]
I'll take a look at this as soon as I can. This is a breaking change, so I'll need to think about it.

**Comment by irony** [comment]
It should be possible to do an arguments constructor to make it backwards compatible. It makes the lib much more usable and faster to be able to save the last state when doing reclustering of new data.

**Comment by harthur** [comment]
Export the `Kmeans` class so you can do `var kmeans = new clusterfck.Kmeans()`, also export `kmeans` so you can use `var clusters = clusterfck.kmeans(vectors)` as usual, and have that just be like:

``` javascript
exports.kmeans = function(vectors, k) {
   return (new Kmeans()).cluster(vectors, k);
}
```

Save kmeans++ initialization for a follow-up pull request. One thing at a time, unless it's necessary for this.

The indentation in the pull request is a mixed bag of tabs and spaces it looks like. This file is 3-space indented (for better or worse), so stick with that for every line.

I'll take a closer look after these things. Thanks for the PR!

**Comment by primaryobjects** [comment]
Done. Added the backwards-compatibility requested and fixed indentation. https://github.com/primaryobjects/clusterfck/commit/d8187230ba2f874826a9ec3bf9c2ad582a3d99a0

The library can now be used two ways:

var clusters = clusterfck.kmeans(colors, 3);

OR

var kmeans = new clusterfck.Kmeans();
var clusters = kmeans.cluster(colors, 3);

File: README.md [hunk 1]
```diff
@@ -26,10 +27,14 @@ var colors = [
    [255, 13, 8]
 ];
 
-var clusters = clusterfck.kmeans(colors, 3);
```
Comment (harthur) [inline README.md:29]: Keep the basic example here, make another example for using it as a stateful object.

File: README.md [hunk 2]
```diff
@@ -39,6 +44,49 @@ The second argument to `kmeans` is the number of clusters you want (default is `
 ]
 ```
 
+# Serialization
```
Comment (harthur) [inline README.md]: This shouldn't be a main header, it should be under the K-means section as it only pertains to kmeans and not hierarchical clustering.

File: README.md [hunk 3]
```diff
@@ -39,6 +44,49 @@ The second argument to `kmeans` is the number of clusters you want (default is `
...
+var json = kmeans.toJSON();
+
+// Deserialize centroids from JSON.
+kmeans = kmeans.fromJSON(json);
+
+// Calculate cluster index from a previously serialized set of centroids.
+var clusterIndex = kmeans.classify([0, 0, 225]);
+```
+
+# Initializing with Existing Centroids
```
Comment (harthur) [inline README.md]: Same for this header and remaining ones, not a main section.

File: lib/kmeans.js [hunk 4]
```diff
@@ -8,28 +12,123 @@ function randomCentroids(points, k) {
    return centroids.slice(0, k);
 }
 
-function closestCentroid(point, centroids, distance) {
+KMeans.prototype.guessK = function(points) {
+    return ~~(Math.sqrt(points.length*0.5));
+}
+  
+KMeans.prototype.initCentroids = function(points, k, distance) {
+   // https://github.com/cmtt/kmeans-js/blob/master/kmeans.js
```
Comment (harthur) [inline lib/kmeans.js:20]: There's no license in that repo, so we can't use it.

File: lib/kmeans.js [hunk 5]
```diff
@@ -8,28 +12,123 @@ function randomCentroids(points, k) {
...
+   var i, cmp1, cmp2;
+   var selectedK = k;
+   var reduce = function(t,c) {var u; for (var i = (v=t[0],1); i < t.length;) v = c(v,t[i],i++,t); i<2 & u && u(); return v;};
+   var addIterator = function (x,y) { return x+y; };
+   var centroids = [];
+
+   /** K-Means++ initialization */
```
Comment (harthur) [inline lib/kmeans.js:27]: Save this for another pull request. Stick to the minimum of changes required for adding a classify method.

File: lib/kmeans.js [hunk 6]
```diff
@@ -8,28 +12,123 @@ function randomCentroids(points, k) {
...
+
+   /** determine the amount of tries */
```
Comment (harthur) [inline lib/kmeans.js:29]: This looks copied from that other repo, can't do that.

File: lib/kmeans.js [hunk 7]
```diff
@@ -8,28 +12,123 @@ function randomCentroids(points, k) {
...
+KMeans.prototype.cluster = function(points, k, distance, snapshotPeriod, snapshotCb) {
+   k = k || this.guessK(points);
 
    distance = distance || "euclidean";
    if (typeof distance == "string") {
       distance = distances[distance];
    }
 
-   var centroids = randomCentroids(points, k);
+   this.centroids = this.initCentroids(points, k, distance); // this.randomCentroids(points, k);
```
Comment (harthur) [inline lib/kmeans.js]: Use `randomCentroids`. Even so, we'd want to take that commented line out, can always look at the history for previous code.

File: lib/kmeans.js (hunk 4 above)
Comment (primaryobjects) [inline lib/kmeans.js:20]: MIT license. https://github.com/cmtt/kmeans-js

**Comment by primaryobjects** [comment]
Ok, I've updated the readme to lessen the font on those sub-sections and to bring back the original code example, with a separate one for instantiating the class. I also put back in the random init for centroids. https://github.com/primaryobjects/clusterfck/commit/6c1a0891ca27cb0b2193cc03a258bd42389e8637

Additional edits to readme https://github.com/primaryobjects/clusterfck/commit/7d3626fa9c3453e669eb0930732dce3872eef926

I left in the code for kmeans++ initialization however, as I think it's a valuable enhancement. The original code has an MIT license. I'll leave it active in my own fork, but feel free to accept this pull and then remove it from the master if you prefer.

Let's get it updated on npm! Thanks!

**Comment by jperl** [comment]
@primaryobjects Thanks for the kmeans++ I was just about to implement that before finding your repo.

**Comment by primaryobjects** [comment]
@jperl I've temporarily put back regular initialization for the pull request for @harthur. Once it's pulled, I'll put the line back in to use kmeans++. Until then, you can grab my fork and just change this line:

https://github.com/primaryobjects/clusterfck/blob/master/lib/kmeans.js#L130

to

``` javascript
// Use kmeans++ initialization.
this.centroids = this.initCentroids(points, k, distance);
```

**Comment by jperl** [comment]
@primaryobjects Gotcha, thanks!

**Comment by irony** [comment]
What's the status on this. I am still using the separate fork - would be great to be able to intitialize clusterfck with centeroids.

**Comment by harthur** [comment]
I merged the code for adding a `classify()` method, and published v0.6.0. Thanks for taking the time to do that primaryobjects.

It's my bad for not making a contribution guide, but in general PRs that are adding multiple features at the same time are harder to review. Even if all the features added are good features (kmeans++ seems like it would be a good addition), you have to determine whether you want all of them at the same time and have to figure out what code goes with what new feature. The complexity makes it easier to put off reviewing it for awhile...

I just marked this repo as unmaintained. You could turn that fork into a new npm package with a new name (retaining the license) and develop from there.

Thanks again.

**Pull request closed without merging**